ga4gh/datamodel/genotype_phenotype.py

```diff
@@ -0,0 +1,338 @@
+        %FILTER%
+        }
+        """
+
+        annotationQuery = annotationQuery.replace("%FILTER%", filter)
+        results = G2PDataset.rdfGraph.query(annotationQuery)
+
+        rows = [row.asdict() for row in results]
+        for row in rows:
+            for k in row:
+                row[k] = row[k].toPython()
+
+        locationQuery = """
+        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+             PREFIX OBO: <http://purl.obolibrary.org/obo/>
+        SELECT distinct *
+        WHERE {
+        ?s    a  OBO:SO_0001059  .
+        ?s ?p ?o .
+        OPTIONAL {?o  rdfs:label ?label .} .
+        %FILTER%
+        }
+        """
+
+        locationRows = []
+        for row in rows:
+            if row['p'] == 'http://purl.org/oban/association_has_subject':
+                location = row['o']
+                locationQuery = locationQuery.replace(
+                    "%FILTER%", "FILTER (?s = <" + location + ">)")
+                # print(locationQuery)
+                results = G2PDataset.rdfGraph.query(locationQuery)
+                locationRows = [row.asdict() for row in results]
+                for row in locationRows:
+                    for k in row:
+                        row[k] = row[k].toPython()
+
+        annotation = self.flatten(rows)
+        location = self.flatten(locationRows)
+        annotation['location'] = location
+        return annotation
+
+    def flatten(self, dict):
+        """
+        Given a dict of dicts,
+        flatten it to a single dict using predicate as keys
+        For multiple occurrences of a predicate, create an array
+        Each value in the dict is an object {val:'x', label:'y'}
+        The value of 's' (subject) is copied to the 'id' property
+        """
+        a = {}
+        for row in dict:
+            obj = {'val': row['o']}
+            if 'label' in row:
+                obj['label'] = row['label']
+
+            if row['p'] in a and \
+                    a[row['p']].__class__.__name__ != "list":
+                asIs = a[row['p']]
+                a[row['p']] = []
+                a[row['p']].append(asIs)
+
+            if row['p'] in a:
+                a[row['p']].append(obj)
+            else:
+                a[row['p']] = obj
+
+            a['id'] = row['s']
+        return a
+
+    def toGA4GH(self, annotation):
+        """
+        given an annotation dict, return a protocol.FeaturePhenotypeAssociation
+        """
+
+        fpa = None
+
+        # annotation keys
+        source = 'http://purl.org/dc/elements/1.1/source'
+        location = 'location'
+        evidenceURI = 'http://purl.obolibrary.org/obo/RO_0002558'
+        hasObject = 'http://purl.org/oban/association_has_object'
+        # location keys
+        GENO_0000408 = 'http://purl.obolibrary.org/obo/GENO_0000408'
+
+        location = annotation['location']
+        if GENO_0000408 in location:
+            id_, ontologySource_ = self.namespaceSplit(
```

Comment:
This id_ here is to prevent the local variable ``id`` from clashing with the global ``id`` function. This is a fairly common Python idiom. No need for ontologySource to end with _.